Automattic/sensei
# Introduce a provider journal store

#### Changes proposed in this Pull Request:

This PR introduces a separate user meta that is going to store a journal for each provider. It includes the functionality that was removed in #3021 and the changes proposed in #2997. 

Initially nothing will be stored in the database. The reason for the PR is to introduce the API and to be available during testing. To enable the functionality the `sensei_enable_journal` filter shoud return true.

The structure of the entry in the database is the following:
```
{
  "course_id: {
    "provider_id": {
      "history": [{
        "timestamp": <timestamp of the change in enrolment>,
        "enrolment_status": true|false|null
      }],
      "log": [{
        "timestamp": <timestamp of the message>,
        "log": "Log message"
      }]
    },
    "another_provider_id": {
      ....
    }
  }
}
```
* A null enrolment status is stored when a provider stops being responsible for the enrolment.
* The actual keys stored are one letter only to save space.

#### Testing instructions
There are no observable changes on the site. The only method that is used at the moment is `register_possible_enrolment_change` and nothing is stored in the database. To enable storing you should return true in `sensei_enable_journal` filter. The rest of the methods can be only verified through the automated tests.

Once the filter is enabled, perform changes to the enrolment of the user and observe that the `sensei_enrolment_providers_journal` entry is updated.
